### PR TITLE
chore: platform design proposal template

### LINKED ISSUE #13752: platform design proposal template
Create a template for use when starting a platform design proposal.

### PR DESCRIPTION
**Pull request by edward-swirldslabs**
**Description**:

Creates a template for platform design proposals. 

**Related issue(s)**:

Fixes #13752

### CODE REVIEW

**Review by imalygin — APPROVED** [review]

File: platform-sdk/docs/proposals/template.md
Comment (anthony-swirldslabs) [inline platform-sdk/docs/proposals/template.md:74]: It would be great to be consistent with whether we mark something as `N/A` or delete inapplicable sections altogether. My personal choice would be the latter to reduce the unnecessary clutter, but I'm good either way as long as we use a consistent approach for every section.

File: platform-sdk/docs/proposals/template.md
Comment (anthony-swirldslabs) [inline platform-sdk/docs/proposals/template.md:76]: This could also include creating new JMH benchmarks for local targeted performance testing, btw.

File: platform-sdk/docs/proposals/template.md
Comment (anthony-swirldslabs) [inline platform-sdk/docs/proposals/template.md:81]: (unrelated to this particular line)
I suggest to also add the optional sections (from the "items to consider") as they may be useful for some designs.

File: platform-sdk/docs/proposals/template.md
Comment (alittley) [inline platform-sdk/docs/proposals/template.md:62]: I think this should be limited to any extra-ordinary unit tests needed. I don't think it's necessary to put "write unit tests for methods" in each proposal.

File: platform-sdk/docs/proposals/template.md
Comment (alittley) [inline platform-sdk/docs/proposals/template.md:64]: What do you mean by unit tests covering "component wiring"? The new pattern for wiring doesn't involve standalone wiring classes anymore

File: platform-sdk/docs/proposals/template.md
Comment (alittley) [inline platform-sdk/docs/proposals/template.md:41]: Perhaps it would be beneficial to note that this section should contain diagrams representing the proposed changes, if applicable

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:74]: I have removed the suggestion to mark N/A and put in the suggestion to remove the section everywhere.

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:76]: I added a mention of JMH benchmarks

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:81]: I've added sections related to the items to consider with text indicating they can be removed.

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:64]: this text has been replaced.

**Review by anthony-swirldslabs — DISMISSED** [review]

File: platform-sdk/docs/proposals/template.md
Comment (poulok) [inline platform-sdk/docs/proposals/template.md:39]: I would give examples of what constitutes a public API:

- Anything defined in protobuf
- Any functional API available for use outside of the module that provides it
- Anything written to disk

File: platform-sdk/docs/proposals/template.md
Comment (poulok) [inline platform-sdk/docs/proposals/template.md:88]: I don't think unit tests, in the traditional sense of the term where only a single method is tested per test, is necessary here. What is more useful, IMO, is describing any higher level functionality that can be covered within the unit test framework. For example, there are various simulators in unit tests that test much more than a single function and are fairly complex, like the consensus unit test framework and the gossip and intake pipeline simulators. These sorts of things need to be planned out and included here.

Comment (poulok) [reply platform-sdk/docs/proposals/template.md:88]: perhaps this is what is meant by `complex component interactions`?

Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:88]: yes that is what I meant by complex component interactions.

Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md]: I've updated the wording.  I still think that the designers should call out any subtle behavior that someone else might be unaware of that needs to be tested.  Let me know if the tweaked wording is acceptable.

**Review by anthony-swirldslabs — DISMISSED** [review]
+1

File: platform-sdk/docs/proposals/template.md
Comment (mxtartaglia-sl) [inline platform-sdk/docs/proposals/template.md:98]:
```suggestion
* Subtle edge cases that might be overlooked.
```

**Review by mxtartaglia-sl — DISMISSED** [review]
Accurate to what has been discussed. Thanks @edward-swirldslabs 
+1

File: platform-sdk/docs/proposals/template.md
Comment (mxtartaglia-sl) [inline platform-sdk/docs/proposals/template.md:1]: Maybe everything should be one level higher and leave title 1 for the title of the proposal?

File: platform-sdk/docs/proposals/template.md
Comment (kfa-aguda) [inline platform-sdk/docs/proposals/template.md:7]: I think we should try to deduplicate as much as possible. Last updated timestamp and committer is readily available in the commit history. I suggest we don't include it here.

File: platform-sdk/docs/proposals/template.md
Comment (mxtartaglia-sl) [inline platform-sdk/docs/proposals/template.md:35]: Proposing a different ordering for this so we can go from general to specific:
## Architecture and/or Components (as components is only meant for wiring framework components)
## Module Organization and repositories
## Core Behavior
## Configuration
## Metrics
## Performance

**Review by alittley — DISMISSED** [review]

File: platform-sdk/docs/proposals/template.md
Comment (poulok) [inline platform-sdk/docs/proposals/template.md:31]: I think this needs more explanation. This is also two sections, in my opinion. The decisions made, and reason why is one. The other is design alternates and why they were not chosen. The notion doc also mentions a design table with criteria and scores.

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:39]: I added the examples you provided.

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:7]: removed

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:35]: reordering to reflect your order here.

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:1]: I've added a title in markdown
```
Title Goes Heres
============
```

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:31]: I've modified this section to reflect your suggested changes.

**Review by alittley — APPROVED** [review]

File: platform-sdk/docs/proposals/template.md
Comment (anthony-swirldslabs) [reply platform-sdk/docs/proposals/template.md:7]: Could you push the commit with the removal please?

File: platform-sdk/docs/proposals/template.md
Comment (mxtartaglia-sl) [reply platform-sdk/docs/proposals/template.md:1]: @edward-swirldslabs I dont see this,
Either way, I think that adds the title as h1 too, sharing the same level as the others.

Comment (alittley) [reply platform-sdk/docs/proposals/template.md:1]: `#` and `=====` are equivalent
`##` and `------` are equivalent

File: platform-sdk/docs/proposals/template.md
Comment (edward-swirldslabs) [reply platform-sdk/docs/proposals/template.md:7]: @anthony-swirldslabs  Just pushed, sorry about that.

**Review by anthony-swirldslabs — APPROVED** [review]

File: platform-sdk/docs/proposals/template.md
Comment (alittley) [reply platform-sdk/docs/proposals/template.md:6]: I think this should still be addressed

**Review by alittley — DISMISSED** [review]

**Review by mxtartaglia-sl — DISMISSED** [review]

**Review by anthony-swirldslabs — APPROVED** [review]

**Review by kfa-aguda — APPROVED** [review]

**Review by anthony-swirldslabs — APPROVED** [review]

**Review by anthony-swirldslabs — APPROVED** [review]
+1

**Review by alittley — APPROVED** [review]
+1

**Comment by imalygin** [comment]
+1

**Review by imalygin — APPROVED** [review]

**Comment by hendrikebbers** [comment]
-1 since I think that more work is needed

**Comment by cody-littley** [comment]
+1

File: platform-sdk/docs/proposals/template.md
Comment (poulok) [inline platform-sdk/docs/proposals/template.md:47]: The notion page we took notes on also included these items:

- Any new metrics or changes to existing metrics
- Changes to default configurations or addition of new configurations

**Review by poulok — APPROVED** [review]
+1 for me. Great job capturing this @edward-swirldslabs!

**Review by rbair23 — APPROVED** [review]

File: platform-sdk/docs/proposals/template.md
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/template.md:11]: We should not user our team names here since they should not be used in public.

File: platform-sdk/docs/proposals/template.md
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/template.md:14]: are the given entries just examples or do you assume that each proposal should have filled out that keys? I assume it makes sense to add a short note to make that clear

File: platform-sdk/docs/proposals/template.md
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/template.md:9]: Once an proposal is merged we can not see in the file if the proposal is already implemented or not. we should add a column for that

File: platform-sdk/docs/proposals/template.md
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/template.md:1]: Will a proposal have a unique id? if yes it should be added here.

File: platform-sdk/docs/proposals/template.md
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/template.md:76]: Question: Do we need to create a proposal whenever that happens? More concrete: If I add api that could be used from outside of a module do I need to add a proposal in future?

File: platform-sdk/docs/proposals/template.md
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/template.md:21]: Where can I define non-goals for the proposal?

File: platform-sdk/docs/proposals/template.md
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/template.md:38]: Does not make sense before the Changes are shown? Often an alternative could be a different api.

File: platform-sdk/docs/proposals/template.md
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/template.md:130]: I would like to have a "Future Work" section that can be used to note what can happen once the proposal is implemented (like the next steps).

File: platform-sdk/docs/proposals/template.md
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/template.md:131]: I would like to have a "Reference" section that can be used to link to additional references

**Comment by hendrikebbers** [comment]
-0.7 since I assume it is too early to create that document and I miss some points

**Pull request merged**
